miner/payload_building.go

```diff
@@ -0,0 +1,160 @@
+	lock := new(sync.Mutex)
+	return &Payload{
+		empty: empty,
+		stop:  make(chan struct{}),
+		lock:  lock,
+		cond:  sync.NewCond(lock),
+	}
+}
+
+// update updates the full-block with latest built version.
+func (payload *Payload) update(block *types.Block) {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	select {
+	case <-payload.stop:
+		return // reject stale update
+	default:
+	}
+	payload.full = block
+	payload.cond.Broadcast() // fire signal for notifying full block
+}
+
+// Resolve returns the latest built payload and also terminates the background
+// thread for updating payload. It's safe to be called multiple times.
+func (payload *Payload) Resolve() *beacon.ExecutableDataV1 {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	select {
+	case <-payload.stop:
+	default:
+		close(payload.stop)
+	}
+	if payload.full != nil {
+		return beacon.BlockToExecutableData(payload.full)
+	}
+	return beacon.BlockToExecutableData(payload.empty)
+}
+
+// ResolveEmpty is basically identical to Resolve, but it expects empty block only.
+// It's only used in tests.
+func (payload *Payload) ResolveEmpty() *beacon.ExecutableDataV1 {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	return beacon.BlockToExecutableData(payload.empty)
+}
+
+// ResolveFull is basically identical to Resolve, but it expects full block only.
+// It's only used in tests.
+func (payload *Payload) ResolveFull() *beacon.ExecutableDataV1 {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	if payload.full == nil {
+		select {
+		case <-payload.stop:
+			return nil
+		default:
+		}
+		payload.cond.Wait()
+	}
+	return beacon.BlockToExecutableData(payload.full)
+}
+
+// buildPayload builds the payload according to the provided parameters.
+func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
+	// Build the initial version with no transaction included. It should be fast
+	// enough to run. The empty payload can at least make sure there is something
+	// to deliver for not missing slot.
+	empty, err := w.getSealingBlock(args.Parent, args.Timestamp, args.FeeRecipient, args.Random, true)
+	if err != nil {
+		return nil, err
+	}
+	// Construct a payload object for return.
+	payload := newPayload(empty)
+
+	// Spin up a routine for updating the payload in background. This strategy
+	// can maximum the revenue for including transactions with highest fee.
+	go func() {
+		// Setup the timer for re-building the payload. The initial clock is kept
+		// for triggering process immediately.
+		timer := time.NewTimer(0)
+		defer timer.Stop()
+
+		// Setup the timer for terminating the process if SECONDS_PER_SLOT (12s in
+		// the Mainnet configuration) have passed since the point in time identified
+		// by the timestamp parameter.
+		endTimer := time.NewTimer(time.Second * 12)
+
+		for {
+			select {
+			case <-timer.C:
+				block, err := w.getSealingBlock(args.Parent, args.Timestamp, args.FeeRecipient, args.Random, false)
+				if err == nil {
```

Comment:
```suggestion
				if err == nil && len(block.Transactions()) > 0 {
```
Maybe? You decide

Reply:
Hmm, I think transaction number is not the best indicator. Personally I feel like we need to use the total transaction fee gathered for validator.

Since the block building can be interrupted internally because of `newpayloadTimeout`, there is no guarantee the re-built block always has higher reward for validator.

I will check it tomorrow if it makes sense to you

Reply:
I have a followup pr that adds the reward to the api, We could do the check there: https://github.com/ethereum/go-ethereum/pull/25998

Reply:
I just added it, please take a look

Reply:
> Hmm, I think transaction number is not the best indicator.

I meant more -- if there are zero transactions, is there any point in calling update? But we could just as well phrase it as "if there are zero fees, is there any point calling update?"

Reply:
Yes, I added the additional check before updating the block
- if full-block is not set yet, set it with the provided one
- if full-block is already set, compare the total transaction fees and only update with higher fee